File: src/org/frankversnel/nl/poortjes/resourceloading/ProcessingShapeLoader.java
package org.frankversnel.nl.poortjes.resourceloading;

import java.io.FileNotFoundException;

import processing.core.PApplet;
import processing.core.PShape;

public class ProcessingShapeLoader extends ResourceLoader<PShape> {
	private PApplet pApplet;

	public ProcessingShapeLoader(PApplet pApplet) {
		this.pApplet = pApplet;
	}

	@Override
	public String loadResource(String fileName) throws FileNotFoundException {
		PShape newShape = pApplet.loadShape(fileName);
		if (newShape == null) {
			throw new FileNotFoundException();
		}

		return addResource(newShape);
	}

}

File: src/main/scala/org/frankversnel/poortjes/resource_loading/processing_shape_loader.scala
package org.frankversnel.poortjes.resource_loading

import processing.core.PApplet;
import processing.core.PShape;

class ProcessingShapeLoader(private val pApplet: PApplet) extends ResourceLoader {
	type T = PShape

	protected def loadResource(filePath: String) : T = {
		val newShape = pApplet.loadShape(filePath);

		return if(newShape != null) {
			newShape
		} else {
            throw new ResourceNotFoundException(filePath)
		}
	}

}
